{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Supplementary Code 1**\n",
    "This notebook was used for designing library for pegRNA abundance screening. For more detail, please read Methods and Supplementary Information.  \n",
    "For detailed information about the completed library, please refer to Supplementary Table 1.\n",
    "\n",
    "Lead contact: Hyoungbum Henry Kim (hkim1@gmail.com)\n",
    "\n",
    "Technical contact: Goosang Yu (gsyu93@gmail.com), Yusang Jung (ys.jung@yuhs.ac)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Directory tree\n",
    "\n",
    "ðŸ“¦Working directory  \n",
    " â”£ ðŸ“‚input  \n",
    " â”ƒ â”— ðŸ“œABL1_lib_design_input_exon4.csv  \n",
    " â”ƒ â”— ðŸ“œABL1_lib_design_input_exon5.csv  \n",
    " â”ƒ â”— ðŸ“œ...  \n",
    " â”— ðŸ“‚output  \n",
    " â”ƒ â”— ðŸ“‚ABL1_exon4  \n",
    " â”ƒ â”ƒ â”— ðŸ“œABL1_ex4_pos1T_A_pe2max_293T.parquet  \n",
    " â”ƒ â”ƒ â”— ðŸ“œABL1_ex4_pos2T_A_pe2max_293T.parquet  \n",
    " â”ƒ â”ƒ â”— ðŸ“œ...  \n",
    " â”— ðŸ“œSuppleCode1.ipynb (this file)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 1: Load reference sequence for input of DeepPrime\n",
    "The pegRNA sequences for creating SNVs in the CDS of each exon of ABL1 and the predicted prime editing efficiency of each pegRNA were generated using DeepPrime. \n",
    "\n",
    "The reference sequences for each input used in DeepPrime are organized in the input directory (example: ABL1_lib_design_input_exon4.csv)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>seq</th>\n",
       "      <th>pos</th>\n",
       "      <th>frame</th>\n",
       "      <th>cds_start</th>\n",
       "      <th>cds_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1_ex4_pos1T_A</td>\n",
       "      <td>GCTCACGTGAGCTCTTTGAGCTTGCCTGTCTCTGTGGGCTGAAGGC...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABL1_ex4_pos2T_A</td>\n",
       "      <td>CTCACGTGAGCTCTTTGAGCTTGCCTGTCTCTGTGGGCTGAAGGCT...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>59</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL1_ex4_pos3C_A</td>\n",
       "      <td>TCACGTGAGCTCTTTGAGCTTGCCTGTCTCTGTGGGCTGAAGGCTG...</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL1_ex4_pos5G_A</td>\n",
       "      <td>ACGTGAGCTCTTTGAGCTTGCCTGTCTCTGTGGGCTGAAGGCTGTT...</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>56</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABL1_ex4_pos6C_A</td>\n",
       "      <td>CGTGAGCTCTTTGAGCTTGCCTGTCTCTGTGGGCTGAAGGCTGTTC...</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>844</th>\n",
       "      <td>ABL1_ex4_pos278G_T</td>\n",
       "      <td>GGAGGTGTACGAGGGCGTGTGGAAGAAATACAGCCTGACGGTGGCC...</td>\n",
       "      <td>278</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>845</th>\n",
       "      <td>ABL1_ex4_pos279G_T</td>\n",
       "      <td>GAGGTGTACGAGGGCGTGTGGAAGAAATACAGCCTGACGGTGGCCG...</td>\n",
       "      <td>279</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>846</th>\n",
       "      <td>ABL1_ex4_pos281A_T</td>\n",
       "      <td>GGTGTACGAGGGCGTGTGGAAGAAATACAGCCTGACGGTGGCCGTG...</td>\n",
       "      <td>281</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847</th>\n",
       "      <td>ABL1_ex4_pos282G_T</td>\n",
       "      <td>GTGTACGAGGGCGTGTGGAAGAAATACAGCCTGACGGTGGCCGTGA...</td>\n",
       "      <td>282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>848</th>\n",
       "      <td>ABL1_ex4_pos283G_T</td>\n",
       "      <td>TGTACGAGGGCGTGTGGAAGAAATACAGCCTGACGGTGGCCGTGAA...</td>\n",
       "      <td>283</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>849 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ID                                                seq  \\\n",
       "0      ABL1_ex4_pos1T_A  GCTCACGTGAGCTCTTTGAGCTTGCCTGTCTCTGTGGGCTGAAGGC...   \n",
       "1      ABL1_ex4_pos2T_A  CTCACGTGAGCTCTTTGAGCTTGCCTGTCTCTGTGGGCTGAAGGCT...   \n",
       "2      ABL1_ex4_pos3C_A  TCACGTGAGCTCTTTGAGCTTGCCTGTCTCTGTGGGCTGAAGGCTG...   \n",
       "3      ABL1_ex4_pos5G_A  ACGTGAGCTCTTTGAGCTTGCCTGTCTCTGTGGGCTGAAGGCTGTT...   \n",
       "4      ABL1_ex4_pos6C_A  CGTGAGCTCTTTGAGCTTGCCTGTCTCTGTGGGCTGAAGGCTGTTC...   \n",
       "..                  ...                                                ...   \n",
       "844  ABL1_ex4_pos278G_T  GGAGGTGTACGAGGGCGTGTGGAAGAAATACAGCCTGACGGTGGCC...   \n",
       "845  ABL1_ex4_pos279G_T  GAGGTGTACGAGGGCGTGTGGAAGAAATACAGCCTGACGGTGGCCG...   \n",
       "846  ABL1_ex4_pos281A_T  GGTGTACGAGGGCGTGTGGAAGAAATACAGCCTGACGGTGGCCGTG...   \n",
       "847  ABL1_ex4_pos282G_T  GTGTACGAGGGCGTGTGGAAGAAATACAGCCTGACGGTGGCCGTGA...   \n",
       "848  ABL1_ex4_pos283G_T  TGTACGAGGGCGTGTGGAAGAAATACAGCCTGACGGTGGCCGTGAA...   \n",
       "\n",
       "     pos  frame  cds_start  cds_end  \n",
       "0      1      1         60      121  \n",
       "1      2      2         59      121  \n",
       "2      3      0         58      121  \n",
       "3      5      2         56      121  \n",
       "4      6      0         55      121  \n",
       "..   ...    ...        ...      ...  \n",
       "844  278      2          0       65  \n",
       "845  279      0          0       64  \n",
       "846  281      2          0       62  \n",
       "847  282      0          0       61  \n",
       "848  283      1          0       60  \n",
       "\n",
       "[849 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "exon_num = 4\n",
    "\n",
    "deepprime_input = f'input/ABL1_lib_design_input_exon{exon_num}.csv'\n",
    "output_path     = f'output/ABL1_exon{exon_num}'\n",
    "os.makedirs(output_path, exist_ok=True)\n",
    "\n",
    "# Load reference sequence DataFrame\n",
    "df_input = pd.read_csv(deepprime_input)\n",
    "\n",
    "df_input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample ID: ABL1_ex4_pos1T_A\n",
      "Input seq: GCTCACGTGAGCTCTTTGAGCTTGCCTGTCTCTGTGGGCTGAAGGCTGTTCCCTGTTTCC(T/A)TTCAGCTCTACGTCTCCTCCGAGAGCCGCTTCAACACCCTGGCCGAGTTGGTTCATCATCA\n"
     ]
    }
   ],
   "source": [
    "# Each data includes ID and input sequence containing WT / Edited information.\n",
    "example_data = df_input.loc[0]\n",
    "\n",
    "print(f'Sample ID:', example_data.ID)\n",
    "print(f'Input seq:', example_data.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 2: Design pegRNAs and get DeepPrime score\n",
    "Using DeepPrime-FT model (Yu et al., Cell, 2023), we designed all available pegRNAs and get predicted prime editing scores for each SNV as bellow.  \n",
    "\n",
    "For using DeepPrime, we used python package 'GenET' version 0.14.x or higher. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DeepPrime-ABL1 exon4:   0%|                                                 | 0/849 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DeepPrime-ABL1 exon4:   1%|                                      | 10/849 [01:05<1:31:48,  6.57s/it]\n"
     ]
    }
   ],
   "source": [
    "from genet.predict import DeepPrime\n",
    "from tqdm import tqdm\n",
    "\n",
    "for idx in tqdm(df_input.index, total=len(df_input),\n",
    "                desc=f'DeepPrime-ABL1 exon{exon_num}',\n",
    "                ncols=100, ascii = ' =', leave=True):\n",
    "\n",
    "    data = df_input.iloc[idx]\n",
    "\n",
    "    sample_id = data.ID\n",
    "    sequence  = data.seq\n",
    "\n",
    "    pegrna = DeepPrime(sequence, name=sample_id, pbs_min=6, pbs_max=17)\n",
    "    pe2max = pegrna.predict(pe_system='PE2max', cell_type='HEK293T')\n",
    "\n",
    "    pe2max.to_parquet(f'{output_path}/{sample_id}_pe2max_293T.parquet', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 3: Select best pegRNAs for screening\n",
    "Based on DeepPrime-FT score, we selected 6 pegRNAs for each varinat as bellow. After then, we only choose 3 pegRNAs for each variants for final list of library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "\n",
    "\n",
    "def reverse_complement(sSeq):\n",
    "    dict_sBases = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N', 'U': 'U', 'n': '',\n",
    "                   '.': '.', '*': '*', 'a': 't', 'c': 'g', 'g': 'c', 't': 'a'}\n",
    "    list_sSeq = list(sSeq)  # Turns the sequence in to a gigantic list\n",
    "    list_sSeq = [dict_sBases[sBase] for sBase in list_sSeq]\n",
    "    return ''.join(list_sSeq)[::-1]\n",
    "\n",
    "# def END: reverse_complement\n",
    "\n",
    "def make_top_pegrna_list(file_dir, pe_sys):\n",
    "\n",
    "    ## Gather all parquet files containing DeepPrime scores for each pegRNAS in the target directory.\n",
    "    file_list = list(glob('%s/%s/results/*.parquet' % (file_dir, pe_sys)))\n",
    "    df_out = pd.DataFrame()\n",
    "\n",
    "    for f in file_list:\n",
    "\n",
    "        ## 1. sort_values('DeepPrime_score')\n",
    "        df = pd.read_parquet(f)\n",
    "        df = df.sort_values('DeepPrime_score', ascending=False)\n",
    "\n",
    "        ## Remove entries with BsmBI sites in the RT-PBS region.\n",
    "        df = df[~df['Edited74_On'].str.contains('GAGACG', na=False, case=False)]\n",
    "        df = df[~df['Edited74_On'].str.contains('CGTCTC', na=False, case=False)]\n",
    "\n",
    "        ## Remove entries with BsmBI sites in the spacer region.\n",
    "        df['gN19'] = df['WT74_On'][5:24]\n",
    "        df = df[~df['gN19'].str.contains('GAGACG', na=False, case=False)]\n",
    "        df = df[~df['gN19'].str.contains('CGTCTC', na=False, case=False)]\n",
    "\n",
    "        ## 2. Extract only the target as a series and keep only the unique values. \n",
    "        series_target = df.WT74_On.unique()\n",
    "\n",
    "        ## 3. Group the sorted df by Seed target.\n",
    "        grouped_df = df.groupby('WT74_On')\n",
    "\n",
    "        ## 4. From ser_target, select the top 3 for each seed target (top 3 per seed target). \n",
    "        \n",
    "        list_df_temp = []\n",
    "\n",
    "        for idx, trgt in enumerate(series_target):\n",
    "            df_top3_pegrna = grouped_df.get_group(trgt).head(2)\n",
    "            list_df_temp.append(df_top3_pegrna)\n",
    "\n",
    "            if idx == 2: break\n",
    "        \n",
    "        ## 5. Output: Store the selected pegRNA information in new dataframes. Add a rank column within those dataframes.\n",
    "        df_temp = pd.concat(list_df_temp)\n",
    "        df_temp['rank'] = df_temp.DeepPrime_score.rank(ascending=False)\n",
    "        \n",
    "        df_out = pd.concat([df_out, df_temp])\n",
    "\n",
    "    return df_out\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Run ###\n",
    "\n",
    "target_list = ['ABL1_ex4', 'ABL1_ex5', 'ABL1_ex6', 'ABL1_ex7', 'ABL1_ex8', 'ABL1_ex9']\n",
    "pe_sys = ['PE2max']\n",
    "base_dir = os.getcwd()\n",
    "\n",
    "\n",
    "for pe in pe_sys:\n",
    "    list_out = []\n",
    "\n",
    "    for gene in target_list:\n",
    "        file_dir = '%s/output/%s' %(base_dir, gene)\n",
    "\n",
    "        df_out = make_top_pegrna_list(file_dir, pe)\n",
    "        list_out.append(df_out)\n",
    "\n",
    "    df = pd.concat(list_out)\n",
    "    df.to_csv('%s/ABL1_top_pegRNAs_%s_220802_BsmBI_filtered_Top6.csv' % (base_dir, pe), index=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "genet",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "undefined.undefined.undefined"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
