{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Supplementary Code 4**\n",
    "This notebook was used for analysis of NGS reads containing intended prime-editing and synonymous mutation marker. For more detail, please read Methods and Supplementary Information. \n",
    "\n",
    "Lead contact: Hyoungbum Henry Kim (hkim1@gmail.com)\n",
    "\n",
    "Technical contact: Goosang Yu (gsyu93@gmail.com), Yusang Jung (ys.jung@yuhs.ac)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Directory tree\n",
    "\n",
    "📦Working directory  \n",
    " ┣ 📂data  \n",
    " ┃ ┣ 📂NGS_FASTQ_files  \n",
    " ┃ ┣ 📂NGS_frequency_table  \n",
    " ┃ ┃ ┣ 📜C4Bosutinib791.txt  \n",
    " ┃ ┃ ┣ 📜C4Control797.txt  \n",
    " ┃ ┃ ┗ 📜...  \n",
    " ┃ ┣ 📂read_counts  \n",
    " ┃ ┣ 📂statistics  \n",
    " ┃  \n",
    " ┣ 📂src  \n",
    " ┃ ┣ 📜Alignment.py  \n",
    " ┃ ┣ 📜VarCalling.py  \n",
    " ┃  \n",
    " ┣ 📂variants_info  \n",
    " ┃ ┣ 📜ex4_info.csv  \n",
    " ┃ ┣ 📜ex5_info.csv  \n",
    " ┃ ┣ 📜ex6_info.csv  \n",
    " ┃ ┣ 📜ex7_info.csv  \n",
    " ┃ ┣ 📜ex8_info.csv  \n",
    " ┃ ┣ 📜ex9_info.csv  \n",
    " ┃ ┣ 📜invivo_ex4_info.csv  \n",
    " ┃ ┗ 📜invivo_ex9_info.csv  \n",
    " ┃  \n",
    " ┗ 📜SuppleCode4.ipynb (this file)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Requirements\n",
    "- CRISPResso2 (>= 2.x.x)\n",
    "- pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variants calling and make read count file\n",
    "After running CRISPResso, generate the read count file. This is the process of creating the foundational file for all analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from glob import glob\n",
    "\n",
    "from src.Alignment import ABL1VUS\n",
    "from src.VarCalling import make_count_file, read_statistics, combine_data, VariantFilter, VariantScore, Normalizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Info] Read counting: K562PE2_unedit_Exon5: 100%|██████████| 36438/36438 [00:03<00:00, 10306.87it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep1_Asciminib: 100%|██████████| 1126274/1126274 [01:52<00:00, 9996.74it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep1_Bosutinib: 100%|██████████| 1310545/1310545 [02:08<00:00, 10160.25it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep1_Dasatinib: 100%|██████████| 1096266/1096266 [01:48<00:00, 10090.48it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep1_DMSO: 100%|██████████| 1231441/1231441 [02:02<00:00, 10064.29it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep1_Imatinib: 100%|██████████| 1290582/1290582 [02:08<00:00, 10016.11it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep1_Nilotinib: 100%|██████████| 1059960/1059960 [01:47<00:00, 9905.26it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep1_Ponatinib: 100%|██████████| 964463/964463 [01:36<00:00, 9955.14it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep2_Asciminib: 100%|██████████| 1532951/1532951 [02:33<00:00, 9975.36it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep2_Bosutinib: 100%|██████████| 1047010/1047010 [01:44<00:00, 9989.00it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep2_Dasatinib: 100%|██████████| 1152683/1152683 [01:56<00:00, 9882.48it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep2_DMSO: 100%|██████████| 1039115/1039115 [01:43<00:00, 10011.08it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep2_Imatinib: 100%|██████████| 1220395/1220395 [02:02<00:00, 10001.29it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep2_Nilotinib: 100%|██████████| 1379467/1379467 [02:19<00:00, 9900.44it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon4_Rep2_Ponatinib: 100%|██████████| 1131439/1131439 [01:52<00:00, 10086.10it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep1_Asciminib: 100%|██████████| 284597/284597 [00:28<00:00, 9976.82it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep1_Bosutinib: 100%|██████████| 190008/190008 [00:18<00:00, 10106.05it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep1_Dasatinib: 100%|██████████| 234142/234142 [00:22<00:00, 10314.68it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep1_DMSO: 100%|██████████| 243513/243513 [00:25<00:00, 9465.35it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep1_Imatinib: 100%|██████████| 287436/287436 [00:29<00:00, 9680.27it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep1_Nilotinib: 100%|██████████| 256333/256333 [00:26<00:00, 9838.81it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep1_Ponatinib: 100%|██████████| 278063/278063 [00:27<00:00, 10126.57it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep2_Asciminib: 100%|██████████| 272710/272710 [00:28<00:00, 9678.84it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep2_Bosutinib: 100%|██████████| 236708/236708 [00:24<00:00, 9859.95it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep2_Dasatinib: 100%|██████████| 225657/225657 [00:22<00:00, 10035.19it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep2_DMSO: 100%|██████████| 265791/265791 [00:25<00:00, 10257.14it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep2_Imatinib: 100%|██████████| 277963/277963 [00:27<00:00, 9990.56it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep2_Nilotinib: 100%|██████████| 296082/296082 [00:30<00:00, 9726.70it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon5_Rep2_Ponatinib: 100%|██████████| 274884/274884 [00:27<00:00, 9907.80it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep1_Asciminib: 100%|██████████| 384040/384040 [00:37<00:00, 10222.48it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep1_Bosutinib: 100%|██████████| 435058/435058 [00:43<00:00, 9930.72it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep1_Dasatinib: 100%|██████████| 2716509/2716509 [04:36<00:00, 9835.91it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep1_DMSO: 100%|██████████| 321185/321185 [00:32<00:00, 9901.96it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep1_Imatinib: 100%|██████████| 350757/350757 [00:35<00:00, 9771.77it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep1_Nilotinib: 100%|██████████| 336212/336212 [00:34<00:00, 9755.02it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep1_Ponatinib: 100%|██████████| 1517771/1517771 [02:32<00:00, 9953.15it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep2_Asciminib: 100%|██████████| 446664/446664 [00:45<00:00, 9837.97it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep2_Bosutinib: 100%|██████████| 512835/512835 [00:50<00:00, 10073.24it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep2_Dasatinib: 100%|██████████| 437648/437648 [00:43<00:00, 9976.04it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep2_DMSO: 100%|██████████| 308569/308569 [00:29<00:00, 10298.77it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep2_Imatinib: 100%|██████████| 312844/312844 [00:32<00:00, 9572.49it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep2_Nilotinib: 100%|██████████| 309247/309247 [00:32<00:00, 9487.04it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon6_Rep2_Ponatinib: 100%|██████████| 498472/498472 [00:50<00:00, 9869.09it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep1_Asciminib: 100%|██████████| 295285/295285 [00:29<00:00, 9944.96it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep1_Bosutinib: 100%|██████████| 307535/307535 [00:31<00:00, 9918.45it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep1_Dasatinib: 100%|██████████| 279471/279471 [00:27<00:00, 10150.44it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep1_DMSO: 100%|██████████| 269842/269842 [00:27<00:00, 9866.02it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep1_Imatinib: 100%|██████████| 286591/286591 [00:28<00:00, 10105.36it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep1_Nilotinib: 100%|██████████| 274591/274591 [00:26<00:00, 10246.24it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep1_Ponatinib: 100%|██████████| 214263/214263 [00:21<00:00, 9740.57it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep2_Asciminib: 100%|██████████| 258860/258860 [00:25<00:00, 10070.55it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep2_Bosutinib: 100%|██████████| 266911/266911 [00:26<00:00, 10180.07it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep2_Dasatinib: 100%|██████████| 279166/279166 [00:28<00:00, 9781.47it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep2_DMSO: 100%|██████████| 258866/258866 [00:25<00:00, 10049.66it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep2_Imatinib: 100%|██████████| 264447/264447 [00:25<00:00, 10239.35it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep2_Nilotinib: 100%|██████████| 269939/269939 [00:28<00:00, 9633.69it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon7_Rep2_Ponatinib: 100%|██████████| 267912/267912 [00:27<00:00, 9901.03it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep1_Asciminib: 100%|██████████| 265345/265345 [00:26<00:00, 10182.20it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep1_Bosutinib: 100%|██████████| 258301/258301 [00:25<00:00, 10235.26it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep1_Dasatinib: 100%|██████████| 276786/276786 [00:27<00:00, 10141.14it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep1_DMSO: 100%|██████████| 245955/245955 [00:24<00:00, 10052.18it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep1_Imatinib: 100%|██████████| 262176/262176 [00:25<00:00, 10159.07it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep1_Nilotinib: 100%|██████████| 243650/243650 [00:23<00:00, 10269.39it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep1_Ponatinib: 100%|██████████| 257091/257091 [00:26<00:00, 9767.50it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep2_Asciminib: 100%|██████████| 262382/262382 [00:26<00:00, 9908.69it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep2_Bosutinib: 100%|██████████| 268718/268718 [00:26<00:00, 10072.83it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep2_Dasatinib: 100%|██████████| 237887/237887 [00:23<00:00, 10151.71it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep2_DMSO: 100%|██████████| 244360/244360 [00:24<00:00, 9938.37it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep2_Imatinib: 100%|██████████| 235388/235388 [00:23<00:00, 9958.89it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep2_Nilotinib: 100%|██████████| 247051/247051 [00:24<00:00, 10136.34it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon8_Rep2_Ponatinib: 100%|██████████| 239411/239411 [00:24<00:00, 9665.48it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep1_Asciminib: 100%|██████████| 41352/41352 [00:04<00:00, 9752.51it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep1_Bosutinib: 100%|██████████| 42781/42781 [00:04<00:00, 9976.49it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep1_Dasatinib: 100%|██████████| 42007/42007 [00:04<00:00, 10041.81it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep1_DMSO: 100%|██████████| 122031/122031 [00:12<00:00, 10034.79it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep1_Imatinib: 100%|██████████| 33500/33500 [00:03<00:00, 10096.21it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep1_Nilotinib: 100%|██████████| 41305/41305 [00:04<00:00, 10115.35it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep1_Ponatinib: 100%|██████████| 39799/39799 [00:03<00:00, 10186.78it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep2_Asciminib: 100%|██████████| 36974/36974 [00:03<00:00, 10323.60it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep2_Bosutinib: 100%|██████████| 45966/45966 [00:04<00:00, 10088.16it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep2_Dasatinib: 100%|██████████| 38426/38426 [00:03<00:00, 10219.37it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep2_DMSO: 100%|██████████| 118661/118661 [00:12<00:00, 9807.51it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep2_Imatinib: 100%|██████████| 42007/42007 [00:04<00:00, 9420.70it/s] \n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep2_Nilotinib: 100%|██████████| 40880/40880 [00:04<00:00, 10020.60it/s]\n",
      "[Info] Read counting: K562PE4K_HTS_Exon9_Rep2_Ponatinib: 100%|██████████| 48962/48962 [00:04<00:00, 9903.26it/s] \n",
      "[Info] Read counting: K562PE4K_unedit_Exon4: 100%|██████████| 591633/591633 [00:58<00:00, 10092.88it/s]\n",
      "[Info] Read counting: K562PE4K_unedit_Exon5: 100%|██████████| 33839/33839 [00:03<00:00, 10006.56it/s]\n",
      "[Info] Read counting: K562PE4K_unedit_Exon6: 100%|██████████| 352558/352558 [00:35<00:00, 10005.75it/s]\n",
      "[Info] Read counting: K562PE4K_unedit_Exon7: 100%|██████████| 361164/361164 [00:35<00:00, 10194.95it/s]\n",
      "[Info] Read counting: K562PE4K_unedit_Exon8: 100%|██████████| 332840/332840 [00:32<00:00, 10143.37it/s]\n",
      "[Info] Read counting: K562PE4K_unedit_Exon9: 100%|██████████| 31833/31833 [00:03<00:00, 10103.50it/s]\n",
      "[Info] Read counting: KCL22PE4K_unedit_Exon5: 100%|██████████| 45978/45978 [00:04<00:00, 10070.60it/s]\n"
     ]
    }
   ],
   "source": [
    "# Make count files from frequency table\n",
    "\n",
    "freq_tables = glob('data/frequency_table/*.txt')\n",
    "\n",
    "for f in freq_tables:\n",
    "\n",
    "    n_sample = os.path.basename(f).replace('.txt', '')\n",
    "    exon_num = n_sample.split('Exon')[1][0]\n",
    "    ref_info = f'variants_info/ex{exon_num}_info.csv'\n",
    "\n",
    "    \n",
    "    df_cnt = make_count_file(f, ref_info)\n",
    "    df_cnt.to_csv(f'data/read_counts/Count_{n_sample}.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Significance analysis\n",
    "Compare the prime-edited sample with the unedited sample to calculate the odds ratio and Fisher's exact test p-value for each variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis: Count_K562PE4K_HTS_Exon4_Rep1_DMSO\n",
      "Analysis: Count_K562PE4K_HTS_Exon4_Rep2_DMSO\n",
      "Analysis: Count_K562PE4K_HTS_Exon5_Rep1_DMSO\n",
      "Analysis: Count_K562PE4K_HTS_Exon5_Rep2_DMSO\n",
      "Analysis: Count_K562PE4K_HTS_Exon6_Rep1_DMSO\n",
      "Analysis: Count_K562PE4K_HTS_Exon6_Rep2_DMSO\n",
      "Analysis: Count_K562PE4K_HTS_Exon7_Rep1_DMSO\n",
      "Analysis: Count_K562PE4K_HTS_Exon7_Rep2_DMSO\n",
      "Analysis: Count_K562PE4K_HTS_Exon8_Rep1_DMSO\n",
      "Analysis: Count_K562PE4K_HTS_Exon8_Rep2_DMSO\n",
      "Analysis: Count_K562PE4K_HTS_Exon9_Rep1_DMSO\n",
      "Analysis: Count_K562PE4K_HTS_Exon9_Rep2_DMSO\n"
     ]
    }
   ],
   "source": [
    "# Calculate statistics with read count for each variants\n",
    "\n",
    "list_control = [\n",
    "    'K562PE4K_HTS_Exon4_Rep1_DMSO',\n",
    "    'K562PE4K_HTS_Exon4_Rep2_DMSO',\n",
    "    'K562PE4K_HTS_Exon5_Rep1_DMSO',\n",
    "    'K562PE4K_HTS_Exon5_Rep2_DMSO',\n",
    "    'K562PE4K_HTS_Exon6_Rep1_DMSO',\n",
    "    'K562PE4K_HTS_Exon6_Rep2_DMSO',\n",
    "    'K562PE4K_HTS_Exon7_Rep1_DMSO',\n",
    "    'K562PE4K_HTS_Exon7_Rep2_DMSO',\n",
    "    'K562PE4K_HTS_Exon8_Rep1_DMSO',\n",
    "    'K562PE4K_HTS_Exon8_Rep2_DMSO',\n",
    "    'K562PE4K_HTS_Exon9_Rep1_DMSO',\n",
    "    'K562PE4K_HTS_Exon9_Rep2_DMSO',\n",
    "    ]\n",
    "\n",
    "for test_sample in list_control:\n",
    "    \n",
    "    background_sample = test_sample.split('_Rep')[0].replace('HTS', 'unedit')\n",
    "\n",
    "    test_file       = f'data/read_counts/Count_{test_sample}.csv'\n",
    "    background_file = f'data/read_counts/Count_{background_sample}.csv'\n",
    "\n",
    "    df_stats = read_statistics(test_file, background_file)\n",
    "\n",
    "    df_stats.to_csv(f'data/statistics/Stat_{test_sample}.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Day 0 vs Day 10 (DMSO treated) analysis\n",
    "Analysis to verify proliferation effects of synonymous/nonsense mutations before assessing drug resistance\n",
    "\n",
    "- Test: After creating variants using prime editing for 20 days, treat with DMSO (negative TKI control) for 10 days.\n",
    "- Control: After creating variants using prime editing for 20 days, immediately harvest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering and Normalization: K562PE4K_HTS_Exon5\n",
      "Filtering and Normalization: K562PE4K_HTS_Exon6\n",
      "Filtering and Normalization: K562PE4K_HTS_Exon7\n",
      "Filtering and Normalization: K562PE4K_HTS_Exon8\n",
      "Filtering and Normalization: K562PE4K_HTS_Exon9\n",
      "Combine: K562PE4K_HTS-Day0_vs_DMSO-Replicate1-Normalization 0.15\n",
      "Combine: K562PE4K_HTS-Day0_vs_DMSO-Replicate2-Normalization 0.15\n",
      "Combine: K562PE4K_HTS-Day0_vs_DMSO-Replicate1-Normalization 0.2\n",
      "Combine: K562PE4K_HTS-Day0_vs_DMSO-Replicate2-Normalization 0.2\n",
      "Combine: K562PE4K_HTS-Day0_vs_DMSO-Replicate1-Normalization 0.3\n",
      "Combine: K562PE4K_HTS-Day0_vs_DMSO-Replicate2-Normalization 0.3\n",
      "Combine: K562PE4K_HTS-Day0_vs_DMSO-Replicate1-Normalization 0.4\n",
      "Combine: K562PE4K_HTS-Day0_vs_DMSO-Replicate2-Normalization 0.4\n",
      "Combine: K562PE4K_HTS-Day0_vs_DMSO-Replicate1-Normalization zscore\n",
      "Combine: K562PE4K_HTS-Day0_vs_DMSO-Replicate2-Normalization zscore\n"
     ]
    }
   ],
   "source": [
    "# Filtering variants and calculate normalized log 2 fold changes\n",
    "sample_tag = 'K562PE4K_HTS'\n",
    "\n",
    "list_sample = [\n",
    "    f'{sample_tag}_Exon5',\n",
    "    f'{sample_tag}_Exon6',\n",
    "    f'{sample_tag}_Exon7',\n",
    "    f'{sample_tag}_Exon8',\n",
    "    f'{sample_tag}_Exon9',\n",
    "    ]\n",
    "\n",
    "list_chem = [\n",
    "    'Imatinib', \n",
    "    'Nilotinib', \n",
    "    'Bosutinib', \n",
    "    'Dasatinib', \n",
    "    'Ponatinib', \n",
    "    'Asciminib', \n",
    "]\n",
    "\n",
    "for sample in list_sample:\n",
    "    '''Day 0 sample은 replicate 1/2가 없는 것들이 있음. \n",
    "    Exon5 - Rep2\n",
    "    Exon6 - Rep1 / 2\n",
    "    Exon7 - Rep1\n",
    "    Exon8 - Rep2\n",
    "    Exon9 - Rep1 / 2\n",
    "    '''\n",
    "    print(f'Filtering and Normalization: {sample}')\n",
    "    test_r1 = f'data/read_counts/Count_{sample}_Rep1_day0.csv'\n",
    "    test_r2 = f'data/read_counts/Count_{sample}_Rep2_day0.csv'\n",
    "\n",
    "    control_r1 = f'data/statistics/Stat_{sample}_Rep1_DMSO.csv'\n",
    "    control_r2 = f'data/statistics/Stat_{sample}_Rep2_DMSO.csv'\n",
    "    \n",
    "    df_rep1, df_rep2 = VariantFilter(test_r1, test_r2, control_r1, control_r2).filter(OR_cutoff=2, p_cutoff=0.05, rpm_cutoff=10)\n",
    "\n",
    "    normal = Normalizer()\n",
    "\n",
    "    for lws_frac in [0.15, 0.2, 0.3, 0.4]:\n",
    "\n",
    "        # LOWESS regression normalization: Day0 = control, DMSO = test로 뒤바꿔서 log fold change를 계산\n",
    "        df_nor1 = normal.lowess(df_rep1, frac=lws_frac, control='test', test='control')\n",
    "        df_nor2 = normal.lowess(df_rep2, frac=lws_frac, control='test', test='control')\n",
    "\n",
    "        df_nor1.to_csv(f'data/statistics/Filtered_{lws_frac}_{sample}_Rep1_Day0DMSO.csv', index=False)\n",
    "        df_nor2.to_csv(f'data/statistics/Filtered_{lws_frac}_{sample}_Rep2_Day0DMSO.csv', index=False)\n",
    "    \n",
    "    # Z-score normalization: Day0 = control, DMSO = test로 뒤바꿔서 log fold change를 계산\n",
    "    df_nor1 = normal.zscore(df_rep1, control='test', test='control')\n",
    "    df_nor2 = normal.zscore(df_rep2, control='test', test='control')\n",
    "\n",
    "    df_nor1.to_csv(f'data/statistics/Filtered_zscore_{sample}_Rep1_Day0DMSO.csv', index=False)\n",
    "    df_nor2.to_csv(f'data/statistics/Filtered_zscore_{sample}_Rep2_Day0DMSO.csv', index=False)\n",
    "\n",
    "\n",
    "for nor_option in [0.15, 0.2, 0.3, 0.4, 'zscore']:\n",
    "\n",
    "    # Exon combine\n",
    "    for rep in [1, 2]:\n",
    "        print(f'Combine: {sample_tag}-Day0_vs_DMSO-Replicate{rep}-Normalization {nor_option}')\n",
    "        files = glob(f'data/statistics/Filtered_{nor_option}_{sample_tag}_Exon*_Rep{rep}_Day0DMSO.csv')\n",
    "        df_merged = combine_data(files)\n",
    "        df_merged.to_csv(f'data/statistics/Filtered_{nor_option}_{sample_tag}_AllExons_Rep{rep}_Day0DMSO.csv', index=False)\n",
    "\n",
    "\n",
    "    # Score calculation: Adjusted LFC / Resistance score\n",
    "    score = VariantScore()\n",
    "\n",
    "    rep_1 = f'data/statistics/Filtered_{nor_option}_{sample_tag}_AllExons_Rep1_Day0DMSO.csv'\n",
    "    rep_2 = f'data/statistics/Filtered_{nor_option}_{sample_tag}_AllExons_Rep2_Day0DMSO.csv'\n",
    "\n",
    "    adjus_LFC = score.calculate(rep_1, rep_2, var_type='SNV')\n",
    "    res_score = score.calculate(rep_1, rep_2, var_type='AA')\n",
    "\n",
    "    adjus_LFC.to_csv(f'data/adjusted_LFC/AdjustedLFC_{nor_option}_{sample_tag}_Day0DMSO.csv')\n",
    "    res_score.to_csv(f'data/resistance_score/ResistanceScore_{nor_option}_{sample_tag}_Day0DMSO.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DMSO vs TKI response analysis\n",
    "Analysis for resistance to drugs\n",
    "\n",
    "- Test: Making variants using Prime editing for 20 days, followed by 10 days of TKI treatment.\n",
    "- Control: Making variants using Prime editing for 20 days, followed by 10 days of DMSO treatment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering variants and calculate normalized log 2 fold changes\n",
    "list_sample = [\n",
    "    'K562PE4K_HTS_Exon4',\n",
    "    'K562PE4K_HTS_Exon5',\n",
    "    'K562PE4K_HTS_Exon6',\n",
    "    'K562PE4K_HTS_Exon7',\n",
    "    'K562PE4K_HTS_Exon8',\n",
    "    'K562PE4K_HTS_Exon9',\n",
    "    ]\n",
    "\n",
    "list_chem = [\n",
    "    'Imatinib', \n",
    "    'Nilotinib', \n",
    "    'Bosutinib', \n",
    "    'Dasatinib', \n",
    "    'Ponatinib', \n",
    "    'Asciminib', \n",
    "]\n",
    "\n",
    "lws_frac = 0.15\n",
    "\n",
    "for sample in list_sample:\n",
    "    for chem in list_chem:\n",
    "\n",
    "        test_r1 = f'data/read_counts/Count_{sample}_Rep1_{chem}.csv'\n",
    "        test_r2 = f'data/read_counts/Count_{sample}_Rep2_{chem}.csv'\n",
    "\n",
    "        control_r1 = f'data/statistics/Stat_{sample}_Rep1_DMSO.csv'\n",
    "        control_r2 = f'data/statistics/Stat_{sample}_Rep2_DMSO.csv'\n",
    "        \n",
    "        df_rep1, df_rep2 = VariantFilter(test_r1, test_r2, control_r1, control_r2).filter(OR_cutoff=2, p_cutoff=0.05, rpm_cutoff=10)\n",
    "\n",
    "        normal = Normalizer()\n",
    "\n",
    "        # LOWESS regression normalization\n",
    "        df_nor1 = normal.lowess(df_rep1, frac=lws_frac)\n",
    "        df_nor2 = normal.lowess(df_rep2, frac=lws_frac)\n",
    "\n",
    "        df_nor1.to_csv(f'data/statistics/Filtered_{sample}_Rep1_{chem}.csv', index=False)\n",
    "        df_nor2.to_csv(f'data/statistics/Filtered_{sample}_Rep2_{chem}.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine all exons data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Combine: K562PE4K_HTS-Imatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Imatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate1\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate2\n",
      "Combine: K562PE4K_HTS-Imatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Imatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate1\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate2\n",
      "Combine: K562PE4K_HTS-Imatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Imatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate1\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate2\n",
      "Combine: K562PE4K_HTS-Imatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Imatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate1\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate2\n",
      "Combine: K562PE4K_HTS-Imatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Imatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate1\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate2\n",
      "Combine: K562PE4K_HTS-Imatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Imatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Nilotinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Bosutinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Dasatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate1\n",
      "Combine: K562PE4K_HTS-Ponatinib-Replicate2\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate1\n",
      "Combine: K562PE4K_HTS-Asciminib-Replicate2\n"
     ]
    }
   ],
   "source": [
    "# Exon combine\n",
    "\n",
    "sample_tag = 'K562PE4K_HTS'\n",
    "\n",
    "list_chem = [\n",
    "    'Imatinib', \n",
    "    'Nilotinib', \n",
    "    'Bosutinib', \n",
    "    'Dasatinib', \n",
    "    'Ponatinib', \n",
    "    'Asciminib', \n",
    "]\n",
    "\n",
    "# Merge all exons for each replicate and chemical\n",
    "for chem in list_chem:\n",
    "    for rep in [1, 2]:\n",
    "        print(f'Combine: {sample_tag}-{chem}-Replicate{rep}')\n",
    "        files = glob(f'data/statistics/Filtered_K562PE4K_HTS_Exon*_Rep{rep}_{chem}.csv')\n",
    "        df_merged = combine_data(files)\n",
    "        df_merged.to_csv(f'data/statistics/Filtered_{sample_tag}_Rep{rep}_{chem}.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Score calculation: Adjusted LFC / Resistance score\n",
    "score = VariantScore()\n",
    " \n",
    "for chem in list_chem:\n",
    "\n",
    "    rep_1 = f'data/statistics/Filtered_K562PE4K_HTS_AllExons_Rep1_{chem}'\n",
    "    rep_2 = f'data/statistics/Filtered_K562PE4K_HTS_AllExons_Rep2_{chem}'\n",
    "\n",
    "    adjus_LFC = score.calculate(rep_1, rep_2, var_type='SNV')\n",
    "    res_score = score.calculate(rep_1, rep_2, var_type='AA')\n",
    "\n",
    "    adjus_LFC.to_csv(f'data/adjusted_LFC/AdjustedLFC_K562PE4K_{chem}.csv')\n",
    "    res_score.to_csv(f'data/resistance_score/ResistanceScore_K562PE4K_{chem}.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "genet",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
